[WebGPU] Create and render to IOSurfaces from the GPU process

mwyrzykowski opened the pull request:
#### e3d86bfd034bfab0d444070be8e4954dca2d62c7
<pre>
[WebGPU] Create and render to IOSurfaces from the GPU process
<a href="https://bugs.webkit.org/show_bug.cgi?id=247587">https://bugs.webkit.org/show_bug.cgi?id=247587</a>
&lt;rdar://101922518&gt;

Reviewed by Dean Jackson.

Create and render to IOSurfaces from the GPU process.

Send them back to the web process via mach port handles.

* Source/WTF/wtf/PlatformEnable.h:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WTF/wtf/PlatformHave.h:
Remove ENABLE_WEBGPU as it overlapped with HAVE_WEBGPU_IMPLEMENTATION.

* Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
(WebCore::GPUDevice::createSurfaceTexture):
(WebCore::GPUDevice::createSurface):
(WebCore::GPUDevice::createSwapChain):
* Source/WebCore/Modules/WebGPU/GPUDevice.h:
* Source/WebCore/Modules/WebGPU/GPUSurface.cpp: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(WebCore::GPUSurface::label const):
(WebCore::GPUSurface::setLabel):
(WebCore::GPUSurface::destroy):
* Source/WebCore/Modules/WebGPU/GPUSurface.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(WebCore::GPUSurface::create):
(WebCore::GPUSurface::backing):
(WebCore::GPUSurface::backing const):
(WebCore::GPUSurface::GPUSurface):
* Source/WebCore/Modules/WebGPU/GPUSurface.idl: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
* Source/WebCore/Modules/WebGPU/GPUSurfaceDescriptor.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(WebCore::GPUSurfaceDescriptor::convertToBacking const):
* Source/WebCore/Modules/WebGPU/GPUSwapChain.cpp: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(WebCore::GPUSwapChain::label const):
(WebCore::GPUSwapChain::setLabel):
(WebCore::GPUSwapChain::prepareForDisplay):
(WebCore::GPUSwapChain::destroy):
* Source/WebCore/Modules/WebGPU/GPUSwapChain.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(WebCore::GPUSwapChain::create):
(WebCore::GPUSwapChain::backing):
(WebCore::GPUSwapChain::backing const):
(WebCore::GPUSwapChain::GPUSwapChain):
* Source/WebCore/Modules/WebGPU/GPUSwapChainDescriptor.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(WebCore::GPUSwapChainDescriptor::convertToBacking const):

Send surface and swap chain concepts from WebProcess layer to GPUProcess layer.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/CMakeLists.txt:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUConvertToBackingContext.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDowncastConvertToBackingContext.cpp:
(PAL::WebGPU::DowncastConvertToBackingContext::convertToBacking):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDowncastConvertToBackingContext.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUSurfaceImpl.cpp: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(PAL::WebGPU::SurfaceImpl::SurfaceImpl):
(PAL::WebGPU::SurfaceImpl::~SurfaceImpl):
(PAL::WebGPU::SurfaceImpl::destroy):
(PAL::WebGPU::SurfaceImpl::setLabelInternal):
(PAL::WebGPU::SurfaceImpl::displayBufferHandle const):
(PAL::WebGPU::SurfaceImpl::drawingBuffer const):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUSurfaceImpl.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUSwapChainImpl.cpp: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(PAL::WebGPU::SwapChainImpl::SwapChainImpl):
(PAL::WebGPU::SwapChainImpl::~SwapChainImpl):
(PAL::WebGPU::SwapChainImpl::destroy):
(PAL::WebGPU::SwapChainImpl::prepareForDisplay):
(PAL::WebGPU::SwapChainImpl::setLabelInternal):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUSwapChainImpl.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUDevice.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUSurface.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(PAL::WebGPU::Surface::label const):
(PAL::WebGPU::Surface::setLabel):
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUSurfaceDescriptor.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUSwapChain.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(PAL::WebGPU::SwapChain::label const):
(PAL::WebGPU::SwapChain::setLabel):
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUSwapChainDescriptor.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::getCSSCanvasContext):
Send surface and swap chain concepts from WebProcess layer to GPUProcess layer.

* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Document.idl:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::getContext):
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/html/HTMLCanvasElement.idl:
Use HAVE_WEBGPU_IMPLEMENTATION instead of ENABLE_WEBGPU as the former was added earlier.

* Source/WebCore/html/canvas/GPUCanvasContext.cpp:
(WebCore::GPUCanvasContext::GPUCanvasContext):
(WebCore::GPUCanvasContext::reshape):
(WebCore::GPUCanvasContext::configure):
(WebCore::GPUCanvasContext::createSwapChainIfNeeded):
(WebCore::GPUCanvasContext::getCurrentTexture):
(WebCore::GPUCanvasContext::unconfigure):
(WebCore::GPUCanvasContext::layerContentsDisplayDelegate):
(WebCore::GPUCanvasContext::prepareForDisplay):
* Source/WebCore/html/canvas/GPUCanvasContext.h:
* Source/WebCore/html/canvas/GPUCanvasContext.idl:
* Source/WebCore/platform/graphics/IntSize.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(wgpuDeviceCreateSwapChain):

* Source/WebGPU/WebGPU/Instance.mm:
(WebGPU::Instance::createSurface):
(wgpuGetProcAddress):
(wgpuInstanceCreateSurface): Deleted.
Moved surface creation to Surface.mm

* Source/WebGPU/WebGPU/Surface.h:
(WebGPU::Surface::create):
(WebGPU::Surface::displayBuffer const):
(WebGPU::Surface::drawingBuffer const):
* Source/WebGPU/WebGPU/Surface.mm:
(optionsFor32BitSurface):
(createIOSurface):
(createSurfaceFromDescriptor):
(wgpuSurfaceCocoaCustomSurfaceGetDisplayBuffer):
(wgpuSurfaceCocoaCustomSurfaceGetDrawingBuffer):
(wgpuInstanceCreateSurface):
(WebGPU::Device::createSurface):
(WebGPU::Surface::Surface):
(WebGPU::Surface::getPreferredFormat):
(WebGPU::Surface::nextDrawable):
Modeled after CAMetalLayer

* Source/WebGPU/WebGPU/SwapChain.h:
(WebGPU::SwapChain::create):
* Source/WebGPU/WebGPU/SwapChain.mm:
(WebGPU::Device::createSwapChain):
(WebGPU::SwapChain::SwapChain):
(WebGPU::SwapChain::present):
(wgpuSwapChainGetCurrentTextureView):
(WebGPU::SwapChain::getCurrentTextureView): Deleted.
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Device::createTexture):
* Source/WebGPU/WebGPU/WebGPUExt.h:
Add IOSurface functions.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
Add new files.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
(WebKit::RemoteDevice::createSurface):
(WebKit::RemoteDevice::createSwapChain):
(WebKit::RemoteDevice::createSurfaceTexture):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSurface.cpp: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(WebKit::RemoteSurface::RemoteSurface):
(WebKit::RemoteSurface::stopListeningForIPC):
(WebKit::RemoteSurface::destroy):
(WebKit::RemoteSurface::setLabel):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSurface.h: Added.
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSurface.messages.in: Added.
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSwapChain.cpp: Added.
(WebKit::RemoteSwapChain::RemoteSwapChain):
(WebKit::RemoteSwapChain::stopListeningForIPC):
(WebKit::RemoteSwapChain::destroy):
(WebKit::RemoteSwapChain::setLabel):
(WebKit::RemoteSwapChain::prepareForDisplay):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSwapChain.h: Added.
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSwapChain.messages.in: Added.
* Source/WebKit/GPUProcess/graphics/WebGPU/WebGPUObjectHeap.cpp:
(WebKit::WebGPU::ObjectHeap::addObject):
(WebKit::WebGPU::ObjectHeap::convertSurfaceFromBacking):
(WebKit::WebGPU::ObjectHeap::convertSwapChainFromBacking):
* Source/WebKit/GPUProcess/graphics/WebGPU/WebGPUObjectHeap.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebGPU/WebGPUConvertFromBackingContext.h:
* Source/WebKit/Shared/WebGPU/WebGPUConvertToBackingContext.h:
* Source/WebKit/Shared/WebGPU/WebGPUSurfaceDescriptor.cpp: Added.
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPUSurfaceDescriptor.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
* Source/WebKit/Shared/WebGPU/WebGPUSurfaceDescriptor.serialization.in: Added.
* Source/WebKit/Shared/WebGPU/WebGPUSwapChainDescriptor.cpp: Added.
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPUSwapChainDescriptor.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
* Source/WebKit/Shared/WebGPU/WebGPUSwapChainDescriptor.serialization.in: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp:
(WebKit::WebGPU::RemoteDeviceProxy::createSurface):
(WebKit::WebGPU::RemoteDeviceProxy::createSwapChain):
(WebKit::WebGPU::RemoteDeviceProxy::createTexture):
(WebKit::WebGPU::RemoteDeviceProxy::createSurfaceTexture):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSurfaceProxy.cpp: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(WebKit::WebGPU::RemoteSurfaceProxy::RemoteSurfaceProxy):
(WebKit::WebGPU::RemoteSurfaceProxy::~RemoteSurfaceProxy):
(WebKit::WebGPU::RemoteSurfaceProxy::destroy):
(WebKit::WebGPU::RemoteSurfaceProxy::setLabelInternal):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSurfaceProxy.h: Added.
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSwapChainProxy.cpp: Added.
(WebKit::WebGPU::RemoteSwapChainProxy::RemoteSwapChainProxy):
(WebKit::WebGPU::RemoteSwapChainProxy::~RemoteSwapChainProxy):
(WebKit::WebGPU::RemoteSwapChainProxy::destroy):
(WebKit::WebGPU::RemoteSwapChainProxy::setLabelInternal):
(WebKit::WebGPU::RemoteSwapChainProxy::prepareForDisplay):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSwapChainProxy.h: Added.
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/WebGPUDowncastConvertToBackingContext.cpp:
(WebKit::WebGPU::DowncastConvertToBackingContext::convertToBacking):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/WebGPUDowncastConvertToBackingContext.h:
Send surface and swap chain concepts from WebProcess layer to GPUProcess layer.

Canonical link: <a href="https://commits.webkit.org/256817@main">https://commits.webkit.org/256817@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f38249ce524a9898216f2b12aeb6dc76d0e9389

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96852 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6121 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29954 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106381 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166665 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100831 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6336 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34849 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89247 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103081 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102526 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4743 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83467 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31756 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86584 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88454 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74671 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/87790 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/153 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/83234 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/141 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21361 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28390 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4724 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/4933 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43878 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/85925 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1369 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40658 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19369 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -349,7 +355,20 @@ WGPUShaderModule wgpuDeviceCreateShaderModule(WGPUDevice device, const WGPUShade
+    return WebGPU::releaseToAPI(WebGPU::fromAPI(device).createSwapChain(*descriptor));
+}
+
+WGPU_EXPORT mach_port_t wgpuDeviceGetSwapChain(WGPUDevice device)
```
I'm not sure if `mach_port_t` is appropriate to return or if I should return a different type

grorg replied:
I'm not sure either, but we have something similar already in our WebGL Shared Event extension.

grorg replied:
@kkinnunen-apple any thoughts?

grorg replied:
Or make a WTF version?

mwyrzykowski replied:
We have WTF::MachSendRight which is essentially a retained mach_port_t from what I can tell, but WebGPU is also C-API, the declaration for this function is in:
```
#ifdef __cplusplus
extern "C" {
#endif
...

```
I can place it inside WebGPU::Surface declared as `class Surface : public WGPUSurfaceImpl` and then return a `WGPUSurface` which is typedef'ed as `typedef struct WGPUSurfaceImpl* WGPUSurface;` 

I thought that was a little over-complicated when I just need the mach port, but that is the other option I see

djg replied:
Why not do the simple thing of returning a `mach_port_t` instead of smuggling it inside some other object to then have the API user extract it? cf. https://twitter.com/tvaneerd/status/977257752820113413?s=20&t=Er99xQIr9d3Pq-nGAN4r3g

mwyrzykowski commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
```diff
@@ -322,6 +334,19 @@ void RemoteDevice::setLabel(String&& label)
+    m_backing->prepareForDisplay(WTFMove(completionHandler));
+}
+#else
+void RemoteDevice::prepareForDisplay(CompletionHandler<void()>&&)
```
This PR probably doesn't build on non-Cocoa, will fix that

kkinnunen-apple commented on Source/WebGPU/WebGPU/Surface.mm:
```diff
@@ -29,9 +29,34 @@
+}
+
+Surface::Surface(const WGPUSurfaceDescriptor& descriptor)
+    : m_surface(getSurfaceFromDescriptor(descriptor))
```
So this one should probably not create the iosurface, rather record the properties.
E.g. `wgpuInstanceCreateSurface(300x200)` creates "empty surface" of size 300x200. It doesn't have the backing store by default (unless it's somehow needed by the api.)
The surface should get its contents from the drawing done by the caller via SwapChain present.

mwyrzykowski replied:
Don't we need to create the IOSurface prior to binding the texture as the color attachment for the render pass?

I.e., in this MTLDevice call:
```
- (nullable id <MTLTexture>)newTextureWithDescriptor:(MTLTextureDescriptor *)descriptor iosurface:(IOSurfaceRef)iosurface plane:(NSUInteger)plane API_AVAILABLE(macos(10.11), ios(11.0));
 ```

we pass the IOSurface already created so that when the render pass completes, we can just call `layer.setContents(ioSurface)`.

I could certainly defer creation of the IOSurface to the time `GPUCanvasContext.getCurrentTexture` is called but I'm not sure how to defer the IOSurface creation later than that point

kkinnunen-apple commented on Source/WebGPU/WebGPU/SwapChain.h:
```diff
@@ -25,32 +25,44 @@
+    const Ref<Device> m_device;
+    WGPUSwapChainDescriptor m_descriptor;
+    WGPUSurface m_displayBuffer;
+    WGPUSurface m_backBuffer;
```
So the drawing buffer is likely a IOSurface, which gets submitted to the WGPUSurface upon present.

The nomenclature should probably be either of these, but consistently:
- front buffer - back buffer
- display buffer - drawing buffer

Swapchain has:
- a drawing buffer
- free buffers
- surface

Surface has:
- display buffer

kkinnunen-apple reviewed (commented):
Nice!
I think there's still confusion of WGPUSurface and IOSurface. WRT comment about the buffering thing, it's unclear if you left this confusion as a future fixup intentionally or if it was unintentional. However, you could fix that before landing so that this would not cause amplification of the confusion.

kkinnunen-apple commented on Source/WebGPU/WebGPU/SwapChain.h:
```diff
@@ -25,32 +25,44 @@
+
+    const Ref<Device> m_device;
+    WGPUSwapChainDescriptor m_descriptor;
+    WGPUSurface m_displayBuffer;
```
There's something funky going on with the references here.
If you contrast `m_device` and `m_backBuffer`, they are references to objects implemented at the same software level. So they should probably be:

1. `const Ref<Device> m_device;`
2. `const Ref<Surface> m_surface;`
3. `RetainPtr<IOSurface> m_drawingBuffer;`

Additionally, it seems incorrect to store `WGPUSwapChainDescriptor`. The client uses that type to communicate the properties here. So here we store the properties in types that are "business as usual code", meaning the native C++ types that the implementation uses. So you store what you need in the likes of `IntSize m_size` et al.

mwyrzykowski replied:
Good point, I will change that, and in fact I don't need to store `WGPUSwapChainDescriptor` at all currently

mwyrzykowski replied:
Also, I don't need to store the `Device` either, so I will remove that, but if I did, should I use `DeviceHolderImpl`? It looks like that is what `QueueImpl` does

kkinnunen-apple commented on Source/WebGPU/WebGPU/Surface.h:
```diff
@@ -39,17 +40,24 @@ class Adapter;
 
     WGPUTextureFormat getPreferredFormat(const Adapter&);
 
+    int width() const;
```
If I remember and understand correctly, these are not used since there's no public C API for these?

mwyrzykowski replied:
Correct, these can be removed.

kkinnunen-apple commented on Source/WebGPU/WebGPU/SwapChain.h:
```diff
@@ -25,32 +25,44 @@
-    TextureView* getCurrentTextureView();
     void present();
 
+    int width() const;
```
If I remember and understand correctly, these are not used since there's no public C API for these?

kkinnunen-apple commented on Source/WebGPU/WebGPU/WebGPUExt.h:
```diff
@@ -84,6 +87,11 @@ typedef struct WGPUTextureDescriptorViewFormats {
 
+typedef struct WGPUSurfaceDescriptorFromCocoaCustomSurface {
+    WGPUChainedStruct chain;
+    IOSurfaceRef surface;
```
So this one is not needed if the mode of the custom surface is that it creates the iosurfaces itself.

mwyrzykowski replied:
I re-purposed it as `WGPUTextureDescriptorCocoaCustomSurface` so the surface can pass the IOSurface to texture creation for `-[MTLDevice newTextureWithDescriptor:ioSurface:plane:]`

kkinnunen-apple commented on Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
```diff
@@ -103,7 +107,22 @@ Ref<GPUBuffer> GPUDevice::createBuffer(const GPUBufferDescriptor& bufferDescript
+    return GPUTexture::create(m_backing->createTexture(textureDescriptor.convertToBacking(), nullptr));
+}
+
+Ref<GPUTexture> GPUDevice::createTexture(const GPUTextureDescriptor& textureDescriptor, const GPUSurface& surface)
```
So here you add part of the functionality that I'd imagine you use to implement feature [1], but I cannot find the API part of the implementation.

kkinnunen-apple commented on Source/WebGPU/WebGPU/WebGPUExt.h:
```diff
@@ -188,6 +196,8 @@ WGPU_EXPORT void wgpuAdapterRequestInvalidDeviceWithBlock(WGPUAdapter adapter, W
 WGPU_EXPORT void wgpuBufferMapAsyncWithBlock(WGPUBuffer buffer, WGPUMapModeFlags mode, size_t offset, size_t size, WGPUBufferMapBlockCallback callback);
 WGPU_EXPORT void wgpuDeviceCreateComputePipelineAsyncWithBlock(WGPUDevice device, WGPUComputePipelineDescriptor const * descriptor, WGPUCreateComputePipelineAsyncBlockCallback callback);
 WGPU_EXPORT void wgpuDeviceCreateRenderPipelineAsyncWithBlock(WGPUDevice device, WGPURenderPipelineDescriptor const * descriptor, WGPUCreateRenderPipelineAsyncBlockCallback callback);
+WGPU_EXPORT WGPUTexture  wgpuDeviceCreateIOSurfaceBackedTexture(WGPUDevice device, const WGPUTextureDescriptor* descriptor, IOSurfaceRef ioSurface);
```
This might be useful function, but it's oddly specific in this scenario.
I suppose it's used to implement feature [1], but I don't think this is the way the API is intended to be implemented.

kkinnunen-apple replied:
So you would need either one, but not the other:
a. Add "create texture that has accessible iosurface (e.g. iosurface backing flag)" + "get the iosurface out of the texture if it was created with the iosurface backing flag"
b. Add "swapchain concept", which needs "surface concept" which needs "create texture out of the surface".

mwyrzykowski replied:
I'll remove `wgpuDeviceCreateIOSurfaceBackedTexture` and use the chaining mechanism instead to achieve this

kkinnunen-apple commented on Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.h:
```diff
@@ -64,7 +65,9 @@ class DeviceImpl final : public Device {
 
     Ref<Buffer> createBuffer(const BufferDescriptor&) final;
-    Ref<Texture> createTexture(const TextureDescriptor&) final;
+    Ref<Texture> createTexture(const TextureDescriptor&, const Surface* = nullptr) final;
```
This sort of device action "create texture for this surface" does not make sense.

Device action: "Create texture for this device".
Surface action: "Create texture out of this surface" or "... surface contents".

So if you want to have these in Device, then they could at least have a different name:

- createTexture(desc)
- createSurfaceTexture(desc, surface)

mwyrzykowski replied:
I will add `Device::createSurfaceTexture(desc, surface)` for now but should I make it a member function on Surface instead? E.g., `Surface::createTexture(desc)`? This is certainly possible, I would need to pass the device handle to the Surface since the device is the one that creates textures

kkinnunen-apple commented on Source/WebCore/html/canvas/GPUCanvasContext.cpp:
```diff
@@ -90,15 +96,60 @@ GPUCanvasContext::CanvasType GPUCanvasContext::canvas()
+    };
+
+    markContextChangedAndNotifyCanvasObservers();
+    return m_configuration->device->createTexture(descriptor, *m_surface);
```
So this is the root of your feature [1].
I.e. you want to create writeable texture out of the drawing buffer.

kkinnunen-apple commented on Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.mm:
```diff
@@ -127,7 +179,54 @@ Ref<Texture> DeviceImpl::createTexture(const TextureDescriptor& descriptor)
     };
 
-    return TextureImpl::create(wgpuDeviceCreateTexture(backing(), &backingDescriptor), descriptor.format, descriptor.dimension, m_convertToBackingContext);
+    IOSurfaceRef ioSurface = surface ? static_cast<const SurfaceImpl*>(surface)->surface() : nullptr;
```
I would understand that the function createTexture(descr) is just a normal texture creation function.

I'd imagine you'd need a new function for "create texture that is backed by the surface". Perhaps conceptually the same as the one you add in the idl, gpusurface.createTexture() or something?

mwyrzykowski committed:
[WebGPU] Create and render to IOSurfaces from the GPU process
https://bugs.webkit.org/show_bug.cgi?id=247587
<rdar://101922518>

Reviewed by Dean Jackson.

Create and render to IOSurfaces from the GPU process.

Send them back to the web process via mach port handles.

* Source/WTF/wtf/PlatformEnable.h:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WTF/wtf/PlatformHave.h:
Remove ENABLE_WEBGPU as it overlapped with HAVE_WEBGPU_IMPLEMENTATION.

* Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
(WebCore::GPUDevice::createSurfaceTexture):
(WebCore::GPUDevice::createSurface):
(WebCore::GPUDevice::createSwapChain):
* Source/WebCore/Modules/WebGPU/GPUDevice.h:
* Source/WebCore/Modules/WebGPU/GPUSurface.cpp: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(WebCore::GPUSurface::label const):
(WebCore::GPUSurface::setLabel):
(WebCore::GPUSurface::destroy):
* Source/WebCore/Modules/WebGPU/GPUSurface.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(WebCore::GPUSurface::create):
(WebCore::GPUSurface::backing):
(WebCore::GPUSurface::backing const):
(WebCore::GPUSurface::GPUSurface):
* Source/WebCore/Modules/WebGPU/GPUSurface.idl: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
* Source/WebCore/Modules/WebGPU/GPUSurfaceDescriptor.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(WebCore::GPUSurfaceDescriptor::convertToBacking const):
* Source/WebCore/Modules/WebGPU/GPUSwapChain.cpp: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(WebCore::GPUSwapChain::label const):
(WebCore::GPUSwapChain::setLabel):
(WebCore::GPUSwapChain::prepareForDisplay):
(WebCore::GPUSwapChain::destroy):
* Source/WebCore/Modules/WebGPU/GPUSwapChain.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(WebCore::GPUSwapChain::create):
(WebCore::GPUSwapChain::backing):
(WebCore::GPUSwapChain::backing const):
(WebCore::GPUSwapChain::GPUSwapChain):
* Source/WebCore/Modules/WebGPU/GPUSwapChainDescriptor.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(WebCore::GPUSwapChainDescriptor::convertToBacking const):

Send surface and swap chain concepts from WebProcess layer to GPUProcess layer.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/CMakeLists.txt:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUConvertToBackingContext.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDowncastConvertToBackingContext.cpp:
(PAL::WebGPU::DowncastConvertToBackingContext::convertToBacking):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDowncastConvertToBackingContext.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUSurfaceImpl.cpp: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(PAL::WebGPU::SurfaceImpl::SurfaceImpl):
(PAL::WebGPU::SurfaceImpl::~SurfaceImpl):
(PAL::WebGPU::SurfaceImpl::destroy):
(PAL::WebGPU::SurfaceImpl::setLabelInternal):
(PAL::WebGPU::SurfaceImpl::displayBufferHandle const):
(PAL::WebGPU::SurfaceImpl::drawingBuffer const):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUSurfaceImpl.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUSwapChainImpl.cpp: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(PAL::WebGPU::SwapChainImpl::SwapChainImpl):
(PAL::WebGPU::SwapChainImpl::~SwapChainImpl):
(PAL::WebGPU::SwapChainImpl::destroy):
(PAL::WebGPU::SwapChainImpl::prepareForDisplay):
(PAL::WebGPU::SwapChainImpl::setLabelInternal):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUSwapChainImpl.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUDevice.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUSurface.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(PAL::WebGPU::Surface::label const):
(PAL::WebGPU::Surface::setLabel):
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUSurfaceDescriptor.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUSwapChain.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(PAL::WebGPU::SwapChain::label const):
(PAL::WebGPU::SwapChain::setLabel):
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUSwapChainDescriptor.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::getCSSCanvasContext):
Send surface and swap chain concepts from WebProcess layer to GPUProcess layer.

* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Document.idl:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::getContext):
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/html/HTMLCanvasElement.idl:
Use HAVE_WEBGPU_IMPLEMENTATION instead of ENABLE_WEBGPU as the former was added earlier.

* Source/WebCore/html/canvas/GPUCanvasContext.cpp:
(WebCore::GPUCanvasContext::GPUCanvasContext):
(WebCore::GPUCanvasContext::reshape):
(WebCore::GPUCanvasContext::configure):
(WebCore::GPUCanvasContext::createSwapChainIfNeeded):
(WebCore::GPUCanvasContext::getCurrentTexture):
(WebCore::GPUCanvasContext::unconfigure):
(WebCore::GPUCanvasContext::layerContentsDisplayDelegate):
(WebCore::GPUCanvasContext::prepareForDisplay):
* Source/WebCore/html/canvas/GPUCanvasContext.h:
* Source/WebCore/html/canvas/GPUCanvasContext.idl:
* Source/WebCore/platform/graphics/IntSize.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(wgpuDeviceCreateSwapChain):

* Source/WebGPU/WebGPU/Instance.mm:
(WebGPU::Instance::createSurface):
(wgpuGetProcAddress):
(wgpuInstanceCreateSurface): Deleted.
Moved surface creation to Surface.mm

* Source/WebGPU/WebGPU/Surface.h:
(WebGPU::Surface::create):
(WebGPU::Surface::displayBuffer const):
(WebGPU::Surface::drawingBuffer const):
* Source/WebGPU/WebGPU/Surface.mm:
(optionsFor32BitSurface):
(createIOSurface):
(createSurfaceFromDescriptor):
(wgpuSurfaceCocoaCustomSurfaceGetDisplayBuffer):
(wgpuSurfaceCocoaCustomSurfaceGetDrawingBuffer):
(wgpuInstanceCreateSurface):
(WebGPU::Device::createSurface):
(WebGPU::Surface::Surface):
(WebGPU::Surface::getPreferredFormat):
(WebGPU::Surface::nextDrawable):
Modeled after CAMetalLayer

* Source/WebGPU/WebGPU/SwapChain.h:
(WebGPU::SwapChain::create):
* Source/WebGPU/WebGPU/SwapChain.mm:
(WebGPU::Device::createSwapChain):
(WebGPU::SwapChain::SwapChain):
(WebGPU::SwapChain::present):
(wgpuSwapChainGetCurrentTextureView):
(WebGPU::SwapChain::getCurrentTextureView): Deleted.
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Device::createTexture):
* Source/WebGPU/WebGPU/WebGPUExt.h:
Add IOSurface functions.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
Add new files.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
(WebKit::RemoteDevice::createSurface):
(WebKit::RemoteDevice::createSwapChain):
(WebKit::RemoteDevice::createSurfaceTexture):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSurface.cpp: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(WebKit::RemoteSurface::RemoteSurface):
(WebKit::RemoteSurface::stopListeningForIPC):
(WebKit::RemoteSurface::destroy):
(WebKit::RemoteSurface::setLabel):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSurface.h: Added.
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSurface.messages.in: Added.
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSwapChain.cpp: Added.
(WebKit::RemoteSwapChain::RemoteSwapChain):
(WebKit::RemoteSwapChain::stopListeningForIPC):
(WebKit::RemoteSwapChain::destroy):
(WebKit::RemoteSwapChain::setLabel):
(WebKit::RemoteSwapChain::prepareForDisplay):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSwapChain.h: Added.
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSwapChain.messages.in: Added.
* Source/WebKit/GPUProcess/graphics/WebGPU/WebGPUObjectHeap.cpp:
(WebKit::WebGPU::ObjectHeap::addObject):
(WebKit::WebGPU::ObjectHeap::convertSurfaceFromBacking):
(WebKit::WebGPU::ObjectHeap::convertSwapChainFromBacking):
* Source/WebKit/GPUProcess/graphics/WebGPU/WebGPUObjectHeap.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebGPU/WebGPUConvertFromBackingContext.h:
* Source/WebKit/Shared/WebGPU/WebGPUConvertToBackingContext.h:
* Source/WebKit/Shared/WebGPU/WebGPUSurfaceDescriptor.cpp: Added.
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPUSurfaceDescriptor.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
* Source/WebKit/Shared/WebGPU/WebGPUSurfaceDescriptor.serialization.in: Added.
* Source/WebKit/Shared/WebGPU/WebGPUSwapChainDescriptor.cpp: Added.
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPUSwapChainDescriptor.h: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
* Source/WebKit/Shared/WebGPU/WebGPUSwapChainDescriptor.serialization.in: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp:
(WebKit::WebGPU::RemoteDeviceProxy::createSurface):
(WebKit::WebGPU::RemoteDeviceProxy::createSwapChain):
(WebKit::WebGPU::RemoteDeviceProxy::createTexture):
(WebKit::WebGPU::RemoteDeviceProxy::createSurfaceTexture):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSurfaceProxy.cpp: Copied from Source/WebCore/html/canvas/GPUCanvasContext.idl.
(WebKit::WebGPU::RemoteSurfaceProxy::RemoteSurfaceProxy):
(WebKit::WebGPU::RemoteSurfaceProxy::~RemoteSurfaceProxy):
(WebKit::WebGPU::RemoteSurfaceProxy::destroy):
(WebKit::WebGPU::RemoteSurfaceProxy::setLabelInternal):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSurfaceProxy.h: Added.
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSwapChainProxy.cpp: Added.
(WebKit::WebGPU::RemoteSwapChainProxy::RemoteSwapChainProxy):
(WebKit::WebGPU::RemoteSwapChainProxy::~RemoteSwapChainProxy):
(WebKit::WebGPU::RemoteSwapChainProxy::destroy):
(WebKit::WebGPU::RemoteSwapChainProxy::setLabelInternal):
(WebKit::WebGPU::RemoteSwapChainProxy::prepareForDisplay):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSwapChainProxy.h: Added.
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/WebGPUDowncastConvertToBackingContext.cpp:
(WebKit::WebGPU::DowncastConvertToBackingContext::convertToBacking):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/WebGPUDowncastConvertToBackingContext.h:
Send surface and swap chain concepts from WebProcess layer to GPUProcess layer.

Canonical link: https://commits.webkit.org/256817@main
